pkg/dns/dns.go

```diff
@@ -0,0 +1,417 @@
+	}
+	if e, ok := e.(*kapi.Endpoints); ok {
+		return kd.generateRecordsForHeadlessService(e, service)
+	}
+	return nil
+}
+
+func (kd *KubeDNS) Records(name string, exact bool) ([]skymsg.Service, error) {
+	glog.Infof("Received DNS Request:%s, exact:%v", name, exact)
+	trimmed := strings.TrimRight(name, ".")
+	segments := strings.Split(trimmed, ".")
+	path := reverseArray(segments)
+	if kd.isPodRecord(path) {
+		response, err := kd.getPodRecord(path)
+		if err == nil {
+			return []skymsg.Service{*response}, nil
+		}
+		return nil, err
+	}
+
+	if exact {
+		key := path[len(path)-1]
+		if key == "" {
+			return []skymsg.Service{}, nil
+		}
+		if record, ok := kd.cache.GetEntry(key, path[:len(path)-1]...); ok {
+			return []skymsg.Service{*(record.(*skymsg.Service))}, nil
+		}
+		return nil, etcd.Error{Code: etcd.ErrorCodeKeyNotFound}
+	}
+
+	// tmp, _ := kd.cache.Serialize("")
+	// glog.Infof("Searching path:%q, %v", path, tmp)
+	records := kd.cache.GetValuesForPathWithRegex(path...)
+	retval := []skymsg.Service{}
+	for _, val := range records {
+		retval = append(retval, *(val.(*skymsg.Service)))
+	}
+	glog.Infof("records:%v, retval:%v, path:%v", records, retval, path)
+	if len(retval) == 0 {
+		return nil, etcd.Error{Code: etcd.ErrorCodeKeyNotFound}
+	}
+	return retval, nil
+}
+
+func (kd *KubeDNS) ReverseRecord(name string) (*skymsg.Service, error) {
+	glog.Infof("Received ReverseRecord Request:%s", name)
+
+	segments := strings.Split(strings.TrimRight(name, "."), ".")
+
+	for _, k := range segments {
+		if k == "*" || k == "any" {
+			return nil, fmt.Errorf("reverse can not contain wildcards")
+		}
+	}
+
+	return nil, fmt.Errorf("must be exactly one service record")
+}
+
+// e.g {"local", "cluster", "pod", "default", "10-0-0-1"}
+func (kd *KubeDNS) isPodRecord(path []string) bool {
+	if len(path) != len(kd.domainPath)+3 {
+		return false
+	}
+	if path[len(kd.domainPath)] != "pod" {
+		return false
+	}
+	for _, segment := range path {
+		if segment == "*" {
+			return false
+		}
+	}
+	return true
+}
+
+func (kd *KubeDNS) getPodRecord(path []string) (*skymsg.Service, error) {
+	ipStr := path[len(path)-1]
+	ip := strings.Replace(ipStr, "-", ".", -1)
+	if parsed := net.ParseIP(ip); parsed != nil {
+		msg := &skymsg.Service{
+			Host:     ip,
+			Port:     0,
+			Priority: 10,
+			Weight:   10,
+			Ttl:      30,
+		}
+		return msg, nil
+	}
+	return nil, fmt.Errorf("Invalid IP Address %v", ip)
+}
+
+// Returns record in a format that SkyDNS understands.
+// Also return the hash of the record.
+func getSkyMsg(ip string, port int) (*skymsg.Service, string) {
+	msg := &skymsg.Service{
+		Host:     ip,
+		Port:     port,
+		Priority: 10,
+		Weight:   10,
+		Ttl:      30,
+	}
+	s := fmt.Sprintf("%v", msg)
+	h := fnv.New32a()
+	h.Write([]byte(s))
+	hash := fmt.Sprintf("%x", h.Sum32())
+	glog.Infof("DNS Record:%s, hash:%s", s, hash)
+	return msg, fmt.Sprintf("%x", hash)
+}
+
+func reverseArray(arr []string) []string {
+	for i := 0; i < len(arr)/2; i++ {
```

Comment:
I dont think your suggestion works.
we have to reverse the array, not reverse sort the array..